File: Programs/matlab/mPLDA/posterior2G.m
% Posterior of two 1-D Gaussian
% Note: mvnpdf is very slow. We can speedup computation by writing our own function
%       where the det(sigma2) is pre-computed.
%       Another speed up is to input the 2x2 precision matrix
function posty = posterior2G(ls, lt, pi, mu, sigma)
K = length(pi);
wlh = zeros(K,K);
sigmaSq = sigma.^2;
IsigmaSq = 1./sigmaSq;
Isigma = 1./sigma;
for p = 1:K,
    for q = 1:K,
        mu2 = [mu(p); mu(q)];
		detIsigma = Isigma(p) * Isigma(q);
        preMat = diag([IsigmaSq(p) IsigmaSq(q)]);		
        wlh(p,q) = pi(p)*pi(q)*pdf2D([ls; lt], mu2, preMat, detIsigma);
    end
end
temp = sum(sum(wlh));
assert(temp>0,'Divided by zeros in posterior_y2 of mPLDA_GroupScoring.m');
posty = wlh/temp; 

function p = pdf2D(x, mu, R, detIsigma)
temp = x-mu;
R = diag(R);
%p = 0.159154943 * detIsigma * exp(-0.5*temp'*R*temp);
p = 0.159154943 * detIsigma * exp(-0.5*sum(temp.^2.*R));

File: Programs/python/mPLDA/posterior2G_.py
import numpy as np
def pdf2D(x, mu, R, detIsigma):
    temp = x - mu
    R = np.diag(R)
    p = 0.159154943 * detIsigma * np.exp(-0.5 * np.sum(temp * temp * R))
    return p

def posterior2G(ls, lt, pi, mu, sigma):
    """
     Posterior of two 1-D Gaussian
     Note: mvnpdf is very slow. We can speedup computation by writing our own function
           where the det(sigma2) is pre-computed.
           Another speed up is to input the 2x2 precision matrix
    """

    K = len(pi)
    wlh = np.zeros(shape=(K, K), dtype=np.float64)
    sigmaSq = sigma * sigma
    IsigmaSq = 1 / sigmaSq
    Isigma = 1 / sigma

    for p in range(K):
        for q in range(K):
            mu2 = np.vstack((mu[p, 0], mu[q, 0]))
            detIsigma = Isigma[p, 0] * Isigma[q, 0]
            preMat = np.diag(np.hstack((IsigmaSq[p, 0], IsigmaSq[q, 0])))
            wlh[p, q] = pi[p, 0] * pi[q, 0] * pdf2D(np.vstack((ls, lt)), mu2, preMat, detIsigma)

    temp = np.sum(np.sum(wlh))
    assert temp > 0, 'Divided by zeros in posterior_y2 of mPLDA_GroupScoring.m'
    posty = wlh / temp
    return posty